File: js/day_8.js
'use strict';

import fs from 'fs';

const data = fs.readFileSync('../resources/day8.in', 'utf-8');

const map = data.split('\n').map(
  (line) => line.split('')
).filter(
  (e) => e.length > 0
);

const ANTENNAS = {};

const MAX_X = map[0].length;
const MAX_Y = map.length;

class Point {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }

  toString() {
    return `Point [${this.x}, ${this.y}]`;
  }

  equals(other) {
    return this.x === other.x && this.y === other.y;
  }
}


function isInBound(point) {
  const x = point.x;
  const y = point.y;

  return x >= 0 && x < MAX_X && y >= 0 && y < MAX_Y;
}


function mapAntennas() {
  for (let y = 0; y < map.length; y++) {
    const line = map[y];

    for (let x = 0; x < line.length; x++) {
      const antenna = line[x];

      if (antenna !== '.') {
        if (!ANTENNAS.hasOwnProperty(antenna)) {
          ANTENNAS[antenna] = [];
        }

        ANTENNAS[antenna].push(new Point(x, y));
      }
    }
  }
}

mapAntennas();

function extractAntiNodes(propagateSignal = false) {
  let knownAntinodes = [];

  return Object.entries(ANTENNAS).flatMap((entry) => {
    const [key, values] = entry;

    const antinodes = values.flatMap((point, idx) => {

      return values.slice(idx+1).flatMap((other) => {

        const newPoints = generatePoint(point, other, propagateSignal).filter(
          (np) => knownAntinodes.every((p) => !p.equals(np)) && isInBound(np)
        );

        knownAntinodes = newPoints.concat(knownAntinodes);

        return newPoints;
      });

    }).filter(
      (e) => isInBound(e)
    );

    return antinodes;
  });
}

function generatePoint(point, otherPoint, propagateSignal) {
  const diffX = point.x - otherPoint.x;
  const diffY = point.y - otherPoint.y;

  const points = [
    new Point(point.x + diffX, point.y + diffY),
    new Point(otherPoint.x - diffX, otherPoint.y - diffY)
  ];

  if (!propagateSignal)
    return points;

  let [pa, pb] = points;
  let signal = 0;
  while (signal < MAX_X) {
    pa = new Point(pa.x + diffX, pa.y + diffY)
    pb = new Point(pb.x - diffX, pb.y - diffY)

    points.push(pa, pb);

    signal++;
  }

  points.push(point, otherPoint);

  return points;
}

function partOne() {
  return extractAntiNodes(false).length;
}

function partTwo() {
  return extractAntiNodes(true).length;
}

console.log(`Part 1: ${partOne()}`);
console.log(`Part 2: ${partTwo()}`);

File: python/day_8.py
# -*- coding: utf-8 -*-

from collections import namedtuple

Point = namedtuple('Point', ['x', 'y'])

data = None
with open('../resources/day8.in', 'r') as file:
    data = file.read().split('\n')
    data = [list(s) for s in data]
    data = [d for d in data if len(d) > 0]

MAX_X = len(data[0])
MAX_Y = len(data)


def extract_antennas():
    antennas = {}
    for y, line in enumerate(data):
        for x, element in enumerate(line):
            if element == '.':
                continue

            if element not in antennas:
                antennas[element] = []

            antennas[element].append(Point(x, y))

    return antennas


def generate_points(point, other_point, propagate_signal):
    diff_x = point.x - other_point.x
    diff_y = point.y - other_point.y

    points = [
        Point(point.x + diff_x, point.y + diff_y),
        Point(other_point.x - diff_x, other_point.y - diff_y)
    ]

    if not propagate_signal:
        return points

    pa, pb = points
    signal = 0
    while signal < MAX_X:
        pa = Point(pa.x + diff_x, pa.y + diff_y)
        pb = Point(pb.x - diff_x, pb.y - diff_y)

        points.extend((
            pa, pb
        ))
        signal += 1

    return [point, other_point] + points


def find_antinode(coords, propagate_signal=False):
    antinodes = []
    for idx, point in enumerate(coords):
        for other_point in coords[idx+1:]:
            antinodes.extend(
                generate_points(point, other_point, propagate_signal)
            )

    return antinodes


def in_bound(antinode):
    x, y = (antinode.x, antinode.y)
    return x >= 0 and x < MAX_X and y >= 0 and y < MAX_Y


def find_all_antinodes(antennas, propagate_signal=False):
    result = []
    for key, coords in antennas.items():
        antinodes = find_antinode(coords, propagate_signal)
        result.extend(filter(lambda x: in_bound(x), antinodes))

    return set(result)

def part_one(antennas):
    return len(find_all_antinodes(antennas, propagate_signal=False))

def part_two(antennas):
    return len(find_all_antinodes(antennas, propagate_signal=True))


antennas = extract_antennas()

print(f"Part 1: {part_one(antennas)}")
print(f"Part 2: {part_two(antennas)}")
